File: Rust/Easy/0110_balanced_binary_tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::cmp;

impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        fn dfs(root: &Option<Rc<RefCell<TreeNode>>>) -> (i32, bool) {
            match root {
                Some(node) => {
                    let (left, bl) = dfs(&node.borrow().left);
                    let (right, br) = dfs(&node.borrow().right);
                    (cmp::max(left, right) + 1, (left - right).abs() <= 1 && bl && br) 
                },
                None => (0, true)
            }
        }
        dfs(&root).1
    }
}

File: python/easy/0110_balanced_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(root: Optional[TreeNode]) -> (bool, int):
            if root is None:
                return (True, 0)
            lBalanced, lHeight = dfs(root.left)
            rBalanced, rHeight = dfs(root.right)
            if not lBalanced or not rBalanced:
                return (False, -1)
            return (abs(rHeight - lHeight) <= 1, max(lHeight, rHeight) + 1)
        balanced, height = dfs(root)
        return balanced
